src/pybind/mgr/cephadm/module.py

```diff
@@ -2535,10 +2535,42 @@ def _apply(self, spec: GenericSpec) -> str:
 
         return self._apply_service_spec(cast(ServiceSpec, spec))
 
+    def _get_candidate_hosts(self, placement: PlacementSpec) -> List[str]:
+        """Return a list of candidate hosts according to the placement specification."""
+        all_hosts = self.cache.get_schedulable_hosts()
+        draining_hosts = [dh.hostname for dh in self.cache.get_draining_hosts()]
+        ls = []
+        if placement.hosts:
+            ls = [h.hostname for h in placement.hosts
+                  if h.hostname in placement.hosts and h.hostname not in draining_hosts]
+        elif placement.label:
+            ls = [x.hostname for x in [h for h in all_hosts if placement.label in h.labels]]
+        elif placement.host_pattern:
+            ls = [x for x in placement.filter_matching_hostspecs(all_hosts)]
+        elif (placement.count is not None or placement.count_per_host is not None):
+            ls = [x.hostname for x in all_hosts]
+        return ls
+
+    def _validate_placement_spec(self, spec: PlacementSpec) -> None:
```

Comment:
I think we need a more specific name for this to signal it's for things like tuned profiles and keyrings rather than actual daemons. Technically this is not a generic placement spec validation function as in many cases placements are very much valid if they include `count` or `count_per_host`

Reply:
changed the name